synapse/cores/postgres.py

```diff
@@ -109,4 +110,14 @@ def _initCorQueries(self, table):
         self._q_getjoin_by_in_int = self._prepQuery(getjoin_by_in_int, table)
         self._q_getjoin_by_in_str = self._prepQuery(getjoin_by_in_str, table)
 
+    def _execute(self, cur, q, r, timeout=None):
+        from psycopg2.extensions import QueryCanceledError
 
+        if timeout:
+            cur.execute('set local statement_timeout = %s', [int(timeout * 1000)])
```

Comment:
Also, because nothing sets the timeout back to 0 ( or whatever makes it eternal ) a subsequent query which ends up using this connection from the pool will be subject to the previously set timeout...